docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,446 @@
+- **To add**
+    - **Bootstrap**
+        - Type: `Bootstrap`
+        - Description: Bootstrap is a reference to a local struct which encapsulates fields to configure the Machine’s bootstrapping mechanism.
+    - **InfrastructureRef [required]**
+        - Type: `corev1.TypedLocalObjectReference`
+        - Description: InfrastructureRef is a required reference to a Custom Resource Definition offered by an infrastructure provider.
+    - **Version [optional]**
+        - Type: `string`
+        - Description: Version defines the desired Kubernetes version. This field is meant to be optionally used by bootstrap providers.
+
+```go
+type MachineStatus struct
+```
+
+- **To remove**
+    - **ProviderStatus [optional]**
+        - Type: `*runtime.RawExtension`
+        - Description: ProviderStatus details a Provider-specific status. It is recommended that providers maintain their own versioned API types that should be serialized/deserialized from this field.
+    - **Versions [optional]**  _Superseded by Version_
+        - Type: `*MachineVersionInfo`
+        - Description: Versions specifies the current versions of software on the corresponding Node (if it exists). This is provided for a few reasons [...].
+    - **Phase [optional]**  _Superseded by Phase_
+        - Type: `*string`
+        - Description: Phase represents the current phase of machine actuation. E.g. Pending, Running, Terminating, Failed etc.
+- **To add**
+    - **Phase**
+        - Type: `MachinePhaseType`
+        - Description: Phase represents the current state of a Machine. This field is not meant to be used as a source of truth. Controllers working on Machines should always respect the documented transition conditions to determine the state.
+    - **Version [optional]**
+        - Type: ``*string`
+        - Description: Version specifies the current version of Kubernetes running on the corresponding Node. This is meant to be a means of bubbling up status from the Node to the Machine. It is entirely optional, but useful for end-user UX if it’s present.
+    - **BootstrapPhase [optional]**
+        - Type: `ControllerPhaseType`
+        - Description: The state of the bootstrap provider.
+    - **InfrastructurePhase [optional]**
+        - Type: `ControllerPhaseType`
+        - Description: The state of the infrastructure provider.
+
+```go
+type Bootstrap struct
+```
+- **To add**
+    - **ConfigRef [optional]**
+        - Type: `*corev1.TypedLocalObjectReference`
+        - Description: ConfigRef is a reference to a bootstrap provider-specific resource that holds configuration details. The reference is optional to allow users/operators to specify Bootstrap.Data without the need of a controller.
```

Comment:
Having one reference assumes there's only one bootstrap provider involved. However, consider the following actual use case:
* One controller provides the basic OS configuration (e.g. cloud-init)
* Another controller provides patches for the OS configuration depending on, for example, the machine class, to accommodate some specific workload (e.g. kernel settings).
 
These two configuration could be made by the same controller, but separating these two concerns makes more sense as the generic OS configuration could be used in multiple scenarios and providers.

Therefore, the `ConfigrRef` could be a list of named references (a `map`) so that each controller knows which reference(s) to set/access. 

The need for multiple references could be extended to other stages, like providing infrastructure configuration before actual provisioning of the machine.

Reply:
I don't think this proposal prevents that possibility. I had imagined that being a bootstrap provider problem, something like:

```
kind: Machine
...
  configRef:
    apiVersion: mergable.something/v1
    kind: MergableBootstrapConfig
    name: my-config
---
apiVersion: mergable.something/v1
kind: MergableBootstrapConfig
spec:
  my-base-config:
      apiVersion: kubeadm.whatever/v1
      kind: KubeadmBootstrapConfig
      name: base-config
  my-workload-tuning-config:
      apiVersion: kernel.config/v1
      kind: SomeCustomKernelConfig
      name: workload-tuning
```

with the implementation of whatever controller reads the `MergableBootstrapConfig` resources defining the merge semantics and preserving the contract with the consumer of the `.Data`.

Reply:
Yes, this is possible if ONE provider manage that, but not if two independent providers must collaborate.

Reply:
I don't follow what you mean. My example had three bootstrap providers: Kubeadm, SomeCustomKernel, and a third "merging" provider whose responsibility it is to broker the interaction between the other two. Wouldn't Kubeadm and SomeCustomKernel count as independent?

Reply:
My point is that all these three controllers must agree on working on a common CRD which is not specified as part or CAPI, so there's some dependency between them.

Reply:
Ah, I see that as a feature – the agreement happens in the controller that owns `MergableBootstrapConfig`, which means both that we can have different merge operations for different kinds of configs, and that iteration can happen independently from CAPI's machine controller (though, if a general pattern emerges, nothing prevents CAPI from shipping a default in the future).

Reply:
> Another controller provides patches for the OS configuration depending on, for example, the machine class, to accommodate some specific workload (e.g. kernel settings). 

That's a pretty wild scenario, if you are doing this I'd recommend against it.  The recommended approach would be to pre-bake your images and simplify your bootstrap sequence.

Reply:
> > Another controller provides patches for the OS configuration depending on, for example, the machine class, to accommodate some specific workload (e.g. kernel settings).
> 
> That's a pretty wild scenario, if you are doing this I'd recommend against it. The recommended approach would be to pre-bake your images and simplify your bootstrap sequence.

This is not a wild scenario.  This is essentially what we do.  Provisioning an instance has nothing to do with creating the artifacts necessary to provide to a cloud (such as user-data), nor does it have much to do with taking an action on that instance afterwards.  Assuming the 'preboot' portion is the same controller as the 'postboot' portion is not a good fit.

Reply:
Software provisioning is it's own fractal tool chain.  We should allow extensions to enable, but it's never been a primary design of cluster api since epoch, and should be completely separate from cluster API.

Reply:
I think I'd like to see the simple case working before we deal with the complexity of merging configurations in the way described here.

Reply:
> Software provisioning is it's own fractal tool chain. We should allow extensions to enable, but it's never been a primary design of cluster api since epoch, and should be completely separate from cluster API.

I understand that the value of providing this kind of extension mechanisms is precisely to allow customization to diverse use cases without having to deal with them in CAPI. The bootstrap controllers live outside CAPI.

What I understand CAPI should ensure, within reasonable boundaries of complexity, is to offer an extension mechanism with is flexible enough to handle diverse use cases. In this case, changing from one single element to a map of elements doesn't seams to be utterly complex. 

I agree, however, with @dhellmann that we can iterate over the design.